### ISSUE TITLE
How to avoid using a retry workaround to get fonts rendered correctly?

### ISSUE THREAD

**Issue by danielearwicker** [mozilla/pdf.js#15708]
I have a kind-of working setup (node using canvas package in debian container) that successfully rasterizes a variety of PDF types. However it is based on a hacky workaround, which is basically:

```js
try {
  return await attemptConversion(pdfData, true);
} catch (x) {
  return await attemptConversion(pdfData, false);
}

const CMAP_URL = "node_modules/pdfjs-dist/cmaps/";
const STANDARD_FONT_DATA_URL = "node_modules/pdfjs-dist/standard_fonts/";

async function attemptConversion(pdfData, firstTry) {
  const pdfDoc = await pdfjsLib.getDocument({
    data: pdfData,
    cMapUrl: CMAP_URL,
    cMapPacked: true,
    standardFontDataUrl: STANDARD_FONT_DATA_URL,
    disableFontFace: firstTry
    stopAtErrors: firstTry
  }).promise;

  // loop through pages and render to canvas to get a bitmap...
```

i.e. try rendering with `disableFontFace = true` and `stopAtErrors = true`, if necessary retry with `disableFontFace = false`.

I have to use this workaround because some real-world PDFs only render correctly with `disableFontFace = false` and some only work with `disableFontFace = true`. So there's a matrix of outcomes:

PDF | `disableFontFace = false` | `disableFontFace = true` |
-----|----------------------------|----------------------------|
test1.pdf  | ✔                          | ❌ no text rendered       |
test2.pdf  | ❌ renders block chars   | ✔                          |

So presumably this workaround is broken if the PDF has a mix of font types.

Is there a configuration I can use that will make it use `disableFontFace` appropriately the fly, depending on what will work for each font?

**Comment by danielearwicker** [comment]
That sounds plausible given that some failures will just look wrong. So I'm not talking about the general case, just this case where it is possible to programmatically determine what setting to use - pdf.js throws an exception, so it has spotted the problem.

In cases such as this, where I can already see it can be programmatically determined, can I do anything to select the appropriate method of rendering fonts on a per-font basis?

**Comment by Snuffleupagus** [comment]
> can I do anything to select the appropriate method of rendering fonts on a per-font basis?

No, and introducing a new option (which usually requires adding/maintaining a bunch of code and tests) for a feature that'd only work *sometimes* doesn't really strike me as a good idea unfortunately (since it'd have to work with all the existing options).
